pkg/controller/data_upload_controller.go

```diff
@@ -146,22 +179,55 @@ func (r *DataUploadReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			return ctrl.Result{}, nil
 		}
 
-		exposeParam := r.setupExposeParam(&du)
+		exposeParam := r.setupExposeParam(du)
 
-		if err := ep.Expose(ctx, getOwnerObject(&du), exposeParam); err != nil {
-			return r.errorOut(ctx, &du, err, "error to expose snapshot", log)
-		}
-		log.Info("Snapshot is exposed")
 		// Expose() will trigger to create one pod whose volume is restored by a given volume snapshot,
 		// but the pod maybe is not in the same node of the current controller, so we need to return it here.
 		// And then only the controller who is in the same node could do the rest work.
+		if err := ep.Expose(ctx, getOwnerObject(du), exposeParam); err != nil {
+			if err := r.client.Get(ctx, req.NamespacedName, du); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return ctrl.Result{}, errors.Wrap(err, "getting DataUpload")
+				}
+			}
+			if isDataUploadInFinalState(du) {
+				log.Warnf("expose snapshot with err %v but it may caused by clean up resources in cancel action", err)
+				r.cleanUp(ctx, du, log)
+				return ctrl.Result{}, nil
+			} else {
+				return r.errorOut(ctx, du, err, "error to expose snapshot", log)
+			}
+		}
+
+		log.Info("Snapshot is exposed")
+
+		// we need to get CR again for it may canceled by dataupload controller on other
+		// nodes when doing expose action, if detectd cancel action we need to clear up the internal
+		// resources created by velero during backup.
+		if err := r.client.Get(ctx, req.NamespacedName, du); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debug("Unable to find DataUpload")
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, errors.Wrap(err, "getting DataUpload")
+		}
+
+		// we need to clean up resources as resources created in Expose it later than cancel action
+		// and need to clean up resources again
+		if du.Spec.Cancel {
```

Comment:
The same, check for isDataUploadInFinalState(du)

Reply:
Updated